packages/zk-certificates/contracts/HumanIDSaltRegistry.sol

```diff
@@ -0,0 +1,179 @@
+     * @param zkCert - Data for the SaltLockingZkCert.
+     * @param idHash - Hash identifying the user. It is supposed to be the poseidon Hash of the name, birthday and citizenship.
+     * @param saltHash - Hash of the salt.
+     */
+    function onZkCertIssuance(
+        SaltLockingZkCert calldata zkCert,
+        uint256 idHash,
+        uint256 saltHash
+    ) external {
+        require(
+            msg.sender == zkCertRegistry,
+            'HumanIDSaltRegistry: only zkCertRegistry can call this function'
+        );
+
+        _registerSaltHash(idHash, saltHash);
+
+        _userData[idHash].zkCerts.push(zkCert.zkCertId);
+        _saltLockingZkCerts[zkCert.zkCertId] = zkCert;
+    }
+
+    /**
+     * @notice Inform salt registry about zkCert revocation. This helps resetting the salt before zkCert expiration if needed.
+     * @param zkCertId - Id of the zkCert to be revoked.
+     */
+    function onZkCertRevocation(uint256 zkCertId) external {
+        require(
+            msg.sender == zkCertRegistry,
+            'HumanIDSaltRegistry: only zkCertRegistry can call this function'
+        );
+
+        _saltLockingZkCerts[zkCertId].revoked = true;
+    }
+
+    /**
+     * @notice Lists zkCerts locking the user to a salt hash.
+     * @param idHash - Hash identifying the user. It is supposed to be the poseidon Hash of the name, birthday and citizenship.
+     * @return List of zkCerts blocking the reset. If it is [], the salt can be reset.
+     */
+    function getSaltLockingZkCerts(
+        uint256 idHash
+    ) public view returns (SaltLockingZkCert[] memory) {
+        require(
+            guardianRegistry.isWhitelisted(msg.sender),
+            'HumanIDSaltRegistry: only whitelisted guardians can call this function'
+        );
+        // a technically advanced user could still extract the data from the blockchain. This is just to make it harder so that the average user can not get it from the block explorer.
+
+        // count zkCerts that are still locking the salt
+        uint256 blockingZkCertsCount = 0;
+        for (uint256 i = 0; i < _userData[idHash].zkCerts.length; i++) {
+            uint256 zkCertId = _userData[idHash].zkCerts[i];
+            if (
+                _saltLockingZkCerts[zkCertId].expirationTime >
+                block.timestamp && // check if not expired
+                !_saltLockingZkCerts[zkCertId].revoked && // check if not revoked
+                guardianRegistry.isWhitelisted(
+                    _saltLockingZkCerts[zkCertId].guardian
+                ) // check if guardian is still whitelisted
+            ) {
+                blockingZkCertsCount++;
+            }
+        }
+
+        SaltLockingZkCert[] memory blockingZkCerts = new SaltLockingZkCert[](
+            blockingZkCertsCount
+        );
+        uint256 fillIndex = 0;
+        // go through all zkCerts and check if any of them is blocking
+        for (uint256 i = 0; i < _userData[idHash].zkCerts.length; i++) {
```

Comment:
I find it clunky to call the same loop twice. I understand the need of it as we don't know the number of the blocking ZkCerts before hand and there is not dynamic array in the memory, but cant we just stop the loop at the first found block ZkCert and return a boolean value instead, because it's all we need in resetSalt.

Reply:
I thought about the use case that guardians want to inform a user about all zkKYCs that need to be revoked before they can reset the salt..

Reply:
I see, there is a snap function that displays such zkKYCs